### PR TITLE
Extend current ImageConfig to allow more values to images as well as …

### LINKED ISSUE #532: K8SSAND-1476 ⁃ Revisit default values for images.Image struct
`Image` currently has a rather complex set of rules to determine default values for its fields. It turns out that the defaults for almost all fields can be guessed by Kubernetes, e.g. if no tag and no pull policy is provided, then the former defaults to `latest` and the latter to `Always`. Also, the registry and the repository are also automatically defaulted to `docker.io` and `library` respectively.

This logic is consequently not necessary at our level. We should be able to remove the logic altogether, but we might need to check how definitive image names are computed and that they remain valid.

On a side note, the current struct doesn't accommodate very nicely tags that are actually image hashes. See [here](https://docs.docker.com/engine/reference/commandline/pull/#pull-an-image-by-digest-immutable-identifier) for the format syntax. We may want to improve that too.

## Update

After discussing it, we'd like to revisit how images are referenced and allow customizing registries, pull secrets, ... by using cass-operator's ImageConfig.
This would require to:

- Replace all occurrences of images.Image in our CRDs and use a string instead (this means we'll have to think about upgrades and possibly introduce a conversion webhook)
- See if we can rely on the same ImageConfig entry that is used by cass-operator
- Evolve the format of the ImageConfig CRD to allow setting a dynamic list of images, each with their own pull secrets/registries/...

```
apiVersion: config.k8ssandra.io/v1beta1
kind: ImageConfig
metadata:
  name: image-config
images:
  system-logger:
     image: "k8ssandra/system-logger:latest"
     imageRegistry: ...
     imagePullSecret:
       name: ....
  config-builder:
     image: "datastax/cass-config-builder:1.0.4-ubi7"
  medusa:
     image: "....."
     imageRegistry: ...
     imagePullSecret:
       name: ....
  # Default settings unless overridden at the image level
  imageRegistry: "localhost:5000"
  imagePullPolicy: Always
  imagePullSecret:
    name: my-secret-pull-registry
defaults:
  # Note, postfix is ignored if repository is not set
  cassandra:
    repository: "k8ssandra/cass-management-api"
  dse:
    repository: "datastax/dse-mgmtapi-6_8"
```

┆Issue is synchronized with this [Jira Story](https://datastax.jira.com/browse/K8OP-272) by [Unito](https://www.unito.io)
┆Issue Number: K8OP-272

### LINKED ISSUE #636: Add support for image namespace override
### What is missing?
In some cases, airgap installations require the namespace to be overridden as well for all images. Currently, imageConfig does not support this.

What we need is a feature that modifies for example `k8ssandra/cass-management-api` images to be `internal/cass-management-api`.

### Why is this needed?
This is required by some airgap installations, which bundle all the images to a single namespace and do not retain the original one (so just changing registry will not work).

┆Issue is synchronized with this [Jira Story](https://datastax.jira.com/browse/CASS-8) by [Unito](https://www.unito.io)
┆Issue Number: CASS-8

### PR DESCRIPTION
**Pull request by burmanm**
**What this PR does**:
Part of the implementation of https://github.com/k8ssandra/k8ssandra-operator/issues/532

**Which issue(s) this PR fixes**:
Fixes https://github.com/k8ssandra/k8ssandra-operator/issues/532
Fixes #636 

**Checklist**
- [ ] Changes manually tested
- [x] Automated Tests added/updated
- [ ] Documentation added/updated
- [x] CHANGELOG.md updated (not required for documentation PRs)
- [x] CLA Signed:  [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

**Comment by burmanm** [comment]
PR still missing modifications to ImagePullPolicy for certain imageTypes in the CassandraDatacenter spec..

File: apis/config/v1beta1/imageconfig_types.go [hunk 1]
```diff
@@ -52,26 +57,88 @@ type Images struct {
 
 	DSEVersions map[string]string `json:"dse,omitempty"`
 
-	SystemLogger string `json:"system-logger"`
+	HCDVersions map[string]string `json:"hcd,omitempty"`
```
Comment (adejanovski) [inline apis/config/v1beta1/imageconfig_types.go:60]: **Issue:** I'm running into an issue when trying to set this:

```
apiVersion: config.k8ssandra.io/v1beta1
kind: ImageConfig
metadata:
  name: image-config
images:
  system-logger: "k8ssandra/system-logger:v1.22.4-dev.fb9eb8c-20241021"
  config-builder: "datastax/cass-config-builder:1.0-ubi8"
  k8ssandra-client: "k8ssandra/k8ssandra-client:v0.5.0"
  hcd:
    "1.0.0": "test/hcd:1.0.0-ubi"
  cassandra:
    "4.0.0": "k8ssandra/cassandra-ubi:latest"
```

Fails with the following error:
```
goroutine 1 [running]:
github.com/k8ssandra/cass-operator/apis/config/v1beta1.(*Images).UnmarshalJSON(0x40006a1650, {0x40002b0103, 0x114, 0x17d})
	/workspace/apis/config/v1beta1/imageconfig_types.go:93 +0x2d4
sigs.k8s.io/json/internal/golang/encoding/json.(*decodeState).object(0x400028d040, {0x13d1c80?, 0x40006a0020?, 0x272?})
	/go/pkg/mod/sigs.k8s.io/json@v0.0.0-20221116044647-bc3834ca7abd/internal/golang/encoding/json/decode.go:649 +0x694
sigs.k8s.io/json/internal/golang/encoding/json.(*decodeState).value(0x400028d040, {0x13d1c80?, 0x40006a0020?, 0x6?})
	/go/pkg/mod/sigs.k8s.io/json@v0.0.0-20221116044647-bc3834ca7abd/internal/golang/encoding/json/decode.go:401 +0x40
sigs.k8s.io/json/internal/golang/encoding/json.(*decodeState).object(0x400028d040, {0x13cb2c0?, 0x40006a0000?, 0x3ea6fc?})
	/go/pkg/mod/sigs.k8s.io/json@v0.0.0-20221116044647-bc3834ca7abd/internal/golang/encoding/json/decode.go:867 +0xeac
sigs.k8s.io/json/internal/golang/encoding/json.(*decodeState).value(0x400028d040, {0x13cb2c0?, 0x40006a0000?, 0x3e9d40?})
	/go/pkg/mod/sigs.k8s.io/json@v0.0.0-20221116044647-bc3834ca7abd/internal/golang/encoding/json/decode.go:401 +0x40
sigs.k8s.io/json/internal/golang/encoding/json.(*decodeState).unmarshal(0x400028d040, {0x13cb2c0?, 0x40006a0000?})
	/go/pkg/mod/sigs.k8s.io/json@v0.0.0-20221116044647-bc3834ca7abd/internal/golang/encoding/json/decode.go:188 +0x104
sigs.k8s.io/json/internal/golang/encoding/json.Unmarshal({0x40002b0000, 0x250, 0x280}, {0x13cb2c0, 0x40006a0000}, {0x4000250b90, 0x2, 0x4000250b70?})
	/go/pkg/mod/sigs.k8s.io/json@v0.0.0-20221116044647-bc3834ca7abd/internal/golang/encoding/json/decode.go:113 +0x124
sigs.k8s.io/json.UnmarshalCaseSensitivePreserveInts(...)
	/go/pkg/mod/sigs.k8s.io/json@v0.0.0-20221116044647-bc3834ca7abd/json.go:62
k8s.io/apimachinery/pkg/runtime/serializer/json.(*Serializer).unmarshal(0x16c7028?, {0x16c7258?, 0x40006a0000?}, {0x40002b0000?, 0x40002a18c0?, 0x13?}, {0x400014c000?, 0x7?, 0x4000514eb0?})
	/go/pkg/mod/k8s.io/apimachinery@v0.29.2/pkg/runtime/serializer/json/json.go:258 +0x18c
k8s.io/apimachinery/pkg/runtime/serializer/json.(*Serializer).Decode(0x400060ca00, {0x400014c000, 0x35e, 0x35f}, 0x0, {0x16c7258, 0x40006a0000})
	/go/pkg/mod/k8s.io/apimachinery@v0.29.2/pkg/runtime/serializer/json/json.go:206 +0x50c
k8s.io/apimachinery/pkg/runtime/serializer/recognizer.(*decoder).Decode(0x2302f20?, {0x400014c000, 0x35e, 0x35f}, 0x0, {0x16c7258, 0x40006a0000})
	/go/pkg/mod/k8s.io/apimachinery@v0.29.2/pkg/runtime/serializer/recognizer/recognizer.go:110 +0x19c
k8s.io/apimachinery/pkg/runtime/serializer/versioning.(*codec).Decode(0x400039d680, {0x400014c000?, 0x35e?, 0x35f?}, 0x0?, {0x16c7258, 0x40006a0000})
	/go/pkg/mod/k8s.io/apimachinery@v0.29.2/pkg/runtime/serializer/versioning/versioning.go:139 +0x80
k8s.io/apimachinery/pkg/runtime.DecodeInto({0xffff39640798?, 0x400039d680?}, {0x400014c000?, 0x400022ae00?, 0x3?}, {0x16c7258, 0x40006a0000})
	/go/pkg/mod/k8s.io/apimachinery@v0.29.2/pkg/runtime/codec.go:63 +0x48
github.com/k8ssandra/cass-operator/pkg/images.LoadImageConfig({0x400014c000, 0x35e, 0x35f})
	/workspace/pkg/images/images.go:57 +0xf8
github.com/k8ssandra/cass-operator/pkg/images.ParseImageConfig({0x40002a1320, 0x1a})
	/workspace/pkg/images/images.go:45 +0x2c
main.main()
	/workspace/cmd/main.go:92 +0x350
panic: interface conversion: interface {} is map[string]interface {}, not string
```

Comment (adejanovski) [reply apis/config/v1beta1/imageconfig_types.go:60]: This works though:
```
apiVersion: config.k8ssandra.io/v1beta1
kind: ImageConfig
metadata:
  name: image-config
images:
  system-logger: "k8ssandra/system-logger:v1.22.4-dev.fb9eb8c-20241021"
  config-builder: "datastax/cass-config-builder:1.0-ubi8"
  k8ssandra-client: "k8ssandra/k8ssandra-client:v0.5.0"
  cassandra:
    "4.0.0": "k8ssandra/cassandra-ubi:latest"
```

File: apis/config/v1beta1/imageconfig_types.go [hunk 2]
```diff
@@ -35,11 +34,17 @@ type ImageConfig struct {
 
 	DefaultImages *DefaultImages `json:"defaults,omitempty"`
 
+	ImagePolicy
```
Comment (adejanovski) [inline apis/config/v1beta1/imageconfig_types.go:37]: **Question:** is this missing `json:",inline"`?

Comment (burmanm) [reply apis/config/v1beta1/imageconfig_types.go:37]: No, it doesn't make any difference in this case.

**Comment by adejanovski** [comment]
@burmanm, should we close this? 
We've had some evolutions lately that might cover this already?

**Comment by burmanm** [comment]
You mean merge? We haven't merged any changes to master.

**Comment by adejanovski** [comment]
no, I mean is it still something we want to do?
Recently you've extended the capabilities of the imageConfig, allowing to add new images dynamically and dealing with defaults more efficiently.

**Comment by burmanm** [comment]
> Recently you've extended the capabilities of the imageConfig, allowing to add new images dynamically and dealing with defaults more efficiently.

Yes, in this PR, not anywhere else. We haven't changed imageConfig properties in anyway elsewhere. That includes HCD version overrides or namespace overrides for airgaps. They are all in this PR only.

Do we not want those features anymore? I thought there was a customer request for the override namespace registry.

File: tests/testdata/image_config_parsing_more_options.yaml [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      "6.8.999": "datastax/dse-server-prototype:latest"
+  hcd:
+      "1.0.0": "datastax/hcd:latest"
+  medusa: "k8ssandra/medusa:latest"
+  reaper: "k8ssandra/reaper:latest"
+imageRegistry: "localhost:5000"
+imagePullPolicy: Always
+imagePullSecret:
+  name: my-secret-pull-registry
+imageNamespace: "enterprise"
```
Comment (adejanovski) [inline tests/testdata/image_config_parsing_more_options.yaml:21]: **Todo:** let's make sure we expose these new fields in the Helm chart.

**Review by adejanovski — APPROVED** [review]

**Pull request merged**